src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -405,7 +405,14 @@ protected void InitializeReferences()
             try
             {
                 // Do the scan starting at the requested position
-                Match? match = runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                Match? match = runner.runmatch;
+                runner.runmatch = null; // Reset runmatch after saving a copy.
+
+                if (match != null && !quick)
+                {
+                    match.Text = input.ToString(); // Only box in case a match Object is going to be returned to the user.
```

Comment:
We should not do this. This will make it incredibly expensive.  Effectively, the existing public APIs that work in terms of Match simply aren't compatible with span inputs.

Reply:
So then you suggest to have a separate Run() method for Spans vs Strings? My attempt here was to try to have a single entry point for both, and only box the string inside the match object whenever the Match would be returned to the  user ( which AFAICT is already what happens today).

Reply:
> and only box the string inside the match object whenever the Match would be returned to the user ( which AFAICT is already what happens today)

This isn't boxing.  This is creating a whole new string.  If the input being searched was a 500K chars, this is going to allocate a 1MB string for each match.  The existing implementation allocates a new Match object (and the associated underlying arrays for captures and such), but not a new string.

> So then you suggest to have a separate Run() method for Spans vs Strings?

Yes.

Related, I assume the https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.RegularExpressions/tests/PrecompiledRegexScenarioTest.cs test is failing with the code as you have it?  It's going to be looking to base.runtext for the input, and if I'm reading the code correctly, it always end up reading null because that's now never being set.  Assuming we need to maintain compatibility and ensure that assemblies previously output from .NET Framework's Regex.CompileToAssembly continue to function, we'll need to ensure that those base RegexRunner fields are configured appropriately when the new Scan method isn't overwritten.

So, I expect you'll need to do something along the lines of:
```C#
// Regex
internal Match? Run(string input, ...)
{
    ... // validate inputs, rent runner

    runner.runtext = input;
    runner.otherfieldstobeset = ...;

    runner.Scan((ReadOnlySpan<char>)input, ...);
    Match m = runner.runmatch;
    if (m is not null && !quick)
    {
        m.Text = input;
    }

    runner.runtext = null;
    runner.otherfieldstobeset = default;

    ... // return runner
    return m;
}

internal Match? Run(ReadOnlySpan<char> input, ...)
{
    ... // validate inputs, rent runner

    runner.Scan(input, ...);
    Match m = runner.runmatch;

    ... // return runner
    return m;
}
```
or something along those lines.

Then, we'd also want to tweak what you have for the base virtual Scan.  Right now in this PR it's this:
```C#
protected internal virtual void Scan(Regex regex, ReadOnlySpan<char> text, int textbeg, int textend, int textstart, int prevlen, bool quick, TimeSpan timeout)
{
    Match? match = Scan(regex, text.ToString(), textbeg, textend, textstart, prevlen, quick, timeout);
    runmatch = match;
}
```
but, again, that's going to allocate a string for the entire input every time a match operation is performed.  And while this would only be an issue for CompileToAssembly implementations, it's _so_ bad I think we still need to care.  We could make it so that cost is only paid if you use an old CompileToAssembly implementation with the new span-based methods, by doing something instead like:
```C#
protected internal virtual void Scan(Regex regex, ReadOnlySpan<char> text, int textbeg, int textend, int textstart, int prevlen, bool quick, TimeSpan timeout)
{
    string s = runtext;
    if (!text.Equals(s)) // checks if the span points to exactly the same ref and length
    {
        s = text.ToString();
        // alternatively, we could just throw NotSupportedException here... basically to use the span inputs, you need to be using an implementation that overrides the new span method, which seems fairly reasonable
    }
    Match? match = Scan(regex, s, textbeg, textend, textstart, prevlen, quick, timeout);
    runmatch = match;
}
```

Reply:
> Related, I assume the main/src/libraries/System.Text.RegularExpressions/tests/PrecompiledRegexScenarioTest.cs test is failing with the code as you have it?

Nope, all tests are passing with the code as-is from this branch, I've also rebased to main locally and all tests continue to pass. I suppose you mean that the problematic line here would be the one that actually compares the test string against the match Group value, but this works because I'm saving the Match.Text value (actually in the line where this comment is) which is where the Value is grabbed from.

Reply:
I mean that runner.runtext isn't being set anywhere, is it?  With the code as you have it now, if someone uses that PrecompiledRegexScenarioTest with the new span-based APIs, it doesn't null ref?

Reply:
It shouldn't null ref because for precompiled regex, they will be using our default implementation of the Scan with Spans overload which will call the original string-based Scan:

https://github.com/dotnet/runtime/blob/2df7931bb3174860fcf907c4188fe6dd457db8e5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L91-L95

 which does set runtext:

https://github.com/dotnet/runtime/blob/2df7931bb3174860fcf907c4188fe6dd457db8e5/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L145

 I did that for backwards compatibility so that precompiled regex would still work if they depended on runtext being set. For new Compiled Regex, that virtual Span-based Scan will be overwritten now with a new Scan which won't set the runtext field any longer.

Reply:
Also, FWIW, I don't want it to look like I'm arguing 😄, I see your point and will adapt the code to do as you suggest, I'm just trying to explain why the test is not failing with the code as I have it today.

Reply:
> It shouldn't null ref

Ok, we should ensure we test that.

> text.ToString()

Ouch.  Similar to my earlier feedback, while we don't need to care _a lot_ about Regex.CompileToAssembly implementations running on .NET 7, we do need to care enough that existing usage of the string-based overloads don't allocate/copy the input string for every match operation.